pkg/defaults/catsrcHelpers.go

```diff
@@ -0,0 +1,147 @@
+	// Get CatalogSource on the cluster
+	cluster := &olm.CatalogSource{}
+	err := client.Get(context.TODO(), wrapper.ObjectKey{
+		Name:      def.Name,
+		Namespace: def.Namespace},
+		cluster)
+	if err != nil && !k8sErrors.IsNotFound(err) {
+		logrus.Errorf("[defaults] Error getting CatalogSource %s - %v", def.Name, err)
+		return err
+	}
+
+	if disable {
+		err = ensureCatsrcAbsent(client, def, cluster)
+	} else {
+		err = ensureCatsrcPresent(client, def, cluster)
+	}
+
+	if err != nil {
+		logrus.Errorf("[defaults] Error processing CatalogSource %s - %v", def.Name, err)
+	}
+
+	return err
+}
+
+// ensureCatsrcAbsent ensure that that the default CatalogSource is not present on the cluster
+func ensureCatsrcAbsent(client wrapper.Client, def olm.CatalogSource, cluster *olm.CatalogSource) error {
+	// CatalogSource is not present on the cluster or has been marked for deletion
+	if cluster.Name == "" || !cluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		logrus.Infof("[defaults] CatalogSource %s not present or has been marked for deletion", def.Name)
+		return nil
+	}
+
+	err := client.Delete(context.TODO(), cluster)
+	if err != nil {
+		return err
+	}
+
+	logrus.Infof("[defaults] Deleting CatalogSource %s", def.Name)
+
+	return err
+}
+
+// ensureCatsrcPresent ensure that that the default CatalogSource is present on the cluster
+func ensureCatsrcPresent(client wrapper.Client, def olm.CatalogSource, cluster *olm.CatalogSource) error {
+	// Create if not present or is deleted
+	if cluster.Name == "" || (!cluster.ObjectMeta.DeletionTimestamp.IsZero() && len(cluster.Finalizers) == 0) {
+		err := client.Create(context.TODO(), &def)
+		if err != nil {
+			return err
+		}
+		logrus.Infof("[defaults] Creating CatalogSource %s", def.Name)
+		return nil
+	}
+
+	if AreCatsrcSpecsEqual(&def.Spec, &cluster.Spec) {
+		logrus.Infof("[defaults] CatalogSource %s default and on cluster specs are same", def.Name)
+		return nil
+	}
+
+	// Update if the spec has changed
+	cluster.Spec = def.Spec
+	err := client.Update(context.TODO(), cluster)
+	if err != nil {
+		return err
+	}
+
+	logrus.Infof("[defaults] Restoring CatalogSource %s", def.Name)
+
+	return nil
+}
+
+// AreCatsrcSpecsEqual returns true if the Specs it receives are the same.
+// Otherwise, the function returns false.
+//
+// The function performs a case insensitive comparison of corresponding
+// attributes.
+//
+// If either of the Specs received is nil, then the function returns false.
+func AreCatsrcSpecsEqual(spec1 *olm.CatalogSourceSpec, spec2 *olm.CatalogSourceSpec) bool {
+	if spec1 == nil || spec2 == nil {
+		return false
+	}
+	if strings.EqualFold(string(spec1.SourceType), string(spec2.SourceType)) &&
+		strings.EqualFold(spec1.ConfigMap, spec2.ConfigMap) &&
+		strings.EqualFold(spec1.Address, spec2.Address) &&
+		strings.EqualFold(spec1.DisplayName, spec2.DisplayName) &&
+		strings.EqualFold(spec1.Publisher, spec2.Publisher) &&
+		strings.EqualFold(spec1.Image, spec2.Image) {
```

Comment:
This should probably include the updatestrategy as well (`updateStrategy.pollInterval` IIRC), so that we can update the default polling on these images

Reply:
@ecordell addressed